qadence/ml_tools/train_utils/base_trainer.py

```diff
@@ -0,0 +1,518 @@
+        """
+        Sets the training DataLoader and computes the number of batches.
+
+        Args:
+            dataloader (DataLoader): The DataLoader for training data.
+        """
+        self._validate_dataloader(dataloader, "train")
+        self._train_dataloader = dataloader
+        self.num_training_batches = self._compute_num_batches(dataloader)
+
+    @property
+    def val_dataloader(self) -> DataLoader:
+        """
+        Returns the validation DataLoader, validating its type.
+
+        Returns:
+            DataLoader: The DataLoader for validation data.
+        """
+        return self._val_dataloader
+
+    @val_dataloader.setter
+    def val_dataloader(self, dataloader: DataLoader) -> None:
+        """
+        Sets the validation DataLoader and computes the number of batches.
+
+        Args:
+            dataloader (DataLoader): The DataLoader for validation data.
+        """
+        self._validate_dataloader(dataloader, "val")
+        self._val_dataloader = dataloader
+        self.num_validation_batches = self._compute_num_batches(dataloader)
+
+    @property
+    def test_dataloader(self) -> DataLoader:
+        """
+        Returns the test DataLoader, validating its type.
+
+        Returns:
+            DataLoader: The DataLoader for testing data.
+        """
+        return self._test_dataloader
+
+    @test_dataloader.setter
+    def test_dataloader(self, dataloader: DataLoader) -> None:
+        """
+        Sets the test DataLoader and computes the number of batches.
+
+        Args:
+            dataloader (DataLoader): The DataLoader for testing data.
+        """
+        self._validate_dataloader(dataloader, "test")
+        self._test_dataloader = dataloader
+        self.num_test_batches = self._compute_num_batches(dataloader)
+
+    @property
+    def config(self) -> TrainConfig:
+        """
+        Returns the training configuration.
+
+        Returns:
+            TrainConfig: The configuration object.
+        """
+        return self._config
+
+    @config.setter
+    def config(self, value: TrainConfig) -> None:
+        """
+        Sets the training configuration and initializes callback and config managers.
+
+        Args:
+            value (Optional[TrainConfig]): The configuration object.
+        """
+        if value and not isinstance(value, TrainConfig):
+            raise TypeError("config must be an instance of TrainConfig.")
+        self._config = value
+        self.callback_manager = CallbacksManager(value)
+        self.config_manager = ConfigManager(value)
+
+    @classmethod
+    def set_use_grad(cls, value: bool) -> None:
+        """
+        Sets the global use_grad flag.
+
+        Args:
+            value (bool): Whether to use gradient-based optimization.
+        """
+        if not isinstance(value, bool):
+            raise TypeError("use_grad must be a boolean value.")
+        cls.use_grad = value
```

Comment:
Currently we have:
1.  Contextmanagers (enable_grad_opt/disable_grad_opt) to change use_grad at an object level. 

2. And set_use_grad() to change use_grad at a class level. 
      > This might be useful in case of parallel processing, where we would like to change use_grad for all instances of the class. Also helps avoiding an extra line of code to set use_grad in case we have multiple trainings - where we can set it up for all instances of Trainer. 


However, we can indeed make use_grad a property with a setter method. This would serve the same purpose as the context managers - to change use_grad at a object level. We can keep either the contextmanager or the setter or both. I added the context managers because they are more pythonic - and the confusion a setter method (trainer.use_grad = True) could cause with classmethod (Trainer.set_use_grad())

Should i just add a setter for use_grad as well?

Reply:
I see. Perhaps it makes more sense to me if the two functionalities were inverted. If `enable_grad_opt` was independent from `Trainer`, then inside that context manager, every new instance of `Trainer` would be spawned with `use_grad=True` (perhaps using a global variable). Correspondingly for `disable_grad_opt`.

This would remove the need of `set_use_grad` and we could have `@property` and `@setter.` around `use_grad` to allow modifications at the instance level.

Reply:
I would be opposed to keeping class level context managers. This will cause problems in case of multiprocessing and threading, as the class context will change during the code - and might interfere with other processes. 
Let's keep the context managers `enable_grad_opt` and `disable_grad_opt` at instance level. For convenience, I will further add a setter method for `use_grad` - in case the user wants to change it during run time. 

So the user will be able to do the following. 

```python 
trainer = Trainer(model, None, config)
with trainer.enable_grad_opt(optimizer):
    trainer.fit()

with trainer.diable_grad_opt(ng_optimizer):
    trainer.fit()
```
or 

```python 
trainer = Trainer(model, None, config)
trainer.use_grad = True
trainer.optimizer = optimizer
trainer.fit()

trainer.use_grad = False
trainer.optimizer = ng_optimizer
trainer.fit()
```


For class level changes - let's keep `set_use_grad` intact. this will help reduce the amount of code needed in case a single optimization method is used throughout.

Reply:
Sounds good